### PR TITLE
[BACKPORT-2.1][SPARK-19372][SQL] Fix throwing a Java exception at df.fliter() due to 64KB bytecode size limit

### PR DESCRIPTION
**Pull request by poplav**
### What changes were proposed in this pull request?

This PR is backport of #17087 to Spark 2.1

### How was this patch tested?

Add a test suite into DataFrameSuite

### CODE REVIEW

**Comment by poplav** [comment]
Looking back at this.  I simply cherry picked the commit from the branch, there appears to be more to this backport.

**Comment by gatorsmile** [comment]
This might be too risky to be merged to 2.1.1.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala [hunk 1]
```diff
@@ -123,6 +124,38 @@ object ExternalCatalogUtils {
     }
     escapePathName(col) + "=" + partitionString
   }
+
+  def prunePartitionsByFilter(
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: Is this method used? I think that this PR includes code that are not related to fixing 64KB issue.
I will investigate all of changes later.

Comment (poplav) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: It is used by the `catalyst/catalog/InMemoryCatalog.scala`.  I picked this up from cherry picking your commit, which introduced other issues.  I may have unnecessarily complicated it.  Let me try and remove this and get back to you.  Thanks.

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala]: Thanks, I changed only seven files in #17087. To update 14 files looks too much.

**Comment by poplav** [comment]
@kiszk , I updated the PR to remove the `prunePartionsByFilter` bit.  Please let me know now.

**Comment by kiszk** [comment]
@poplav it looks good
@gatorsmile Do you think it is ok for backport now? The previous commit included unnecessary changes.

**Comment by gatorsmile** [comment]
@poplav This is not a regression from 2.0, right? 

Since we might not release 2.1.2, this PR might not be merged to upstream after a discussion with @zsxwing Maybe you can patch it in your private build.

**Comment by poplav** [comment]
Was this working in 2.0 in the first place?  I want to get this into 2.1.1

**Comment by gatorsmile** [comment]
You can patch it to your forked version.

**Comment by pmishr1** [comment]
@poplav were you able to patch this PR and build successfully on top of 2.1.1 ?

**Comment by poplav** [comment]
@pmishr1 Yeah it worked this PR/branch needs one more commit that is at https://github.com/poplav/spark/commit/48ea442ea670cc612b91ad037b223750aa70b918.  I would update this PR with that commit, but doesn't look like this is going anywhere

**Comment by dongjoon-hyun** [comment]
Gentle ping, @poplav .

**Comment by dongjoon-hyun** [comment]
Ping, @poplav .

**Comment by poplav** [comment]
Was away on vacation.  Closing PR, thanks.

**Comment by dongjoon-hyun** [comment]
Thank you, @poplav ! :D

**Pull request closed without merging**